unsloth/models/llama.py

```diff
@@ -572,100 +580,49 @@ def LlamaAttention_fast_forward(
-        Q = Q.transpose(1, 2)
-        K = K.transpose(1, 2)
-        V = V.transpose(1, 2)
-
-        # Group query attention
-        if n_groups != 1:
-            K = K.view(bsz, kv_seq_len, n_kv_heads, 1, head_dim)
-            V = V.view(bsz, kv_seq_len, n_kv_heads, 1, head_dim)
-            K = K.expand(bsz, kv_seq_len, n_kv_heads, n_groups, head_dim)
-            V = V.expand(bsz, kv_seq_len, n_kv_heads, n_groups, head_dim)
-            if hidden_states.requires_grad:
-                K = K.reshape(bsz, kv_seq_len, n_heads, head_dim)
-                V = V.reshape(bsz, kv_seq_len, n_heads, head_dim)
-            else:
-                Q = Q.view(bsz, q_len, n_kv_heads, n_groups, head_dim)
-        A = xformers_attention(Q, K, V, attn_bias = causal_mask)
-        A = A.view(bsz, q_len, n_heads, head_dim)
-
-    elif HAS_FLASH_ATTENTION and attention_mask is None:
-        Q = Q.transpose(1, 2)
-        K = K.transpose(1, 2)
-        V = V.transpose(1, 2)
-        A = flash_attn_func(Q, K, V, causal = True)
-    else:
-        # when qlen==vlen and attn_mask is None, we should use causal attention
-        Q_len = Q.shape[-2]
-        K_len = K.shape[-2]
-        if attention_mask is None and Q_len == K_len:
-            is_causal = True
-        else:
-            is_causal = False
-        # Grouped query attention
-        if SDPA_HAS_GQA:
-            # Needs (batch_size, n_heads, seq_len, head_dim)
-            # is_casual and attention_mask must not be both set!
-            A = scaled_dot_product_attention(
-                Q,
-                K,
-                V,
-                attn_mask = attention_mask,
-                is_causal = is_causal,
-                enable_gqa = n_groups != 1,
-            )
-            # Go back to (batch_size, seq_len, n_heads, head_dim)
-            A = A.transpose(1, 2)  # .contiguous()
-        else:
-            if n_groups != 1:
-                K = K[:, :, None, :, :].expand(
-                    bsz, n_kv_heads, n_groups, kv_seq_len, head_dim
-                )
-                V = V[:, :, None, :, :].expand(
-                    bsz, n_kv_heads, n_groups, kv_seq_len, head_dim
-                )
-                K = K.reshape(bsz, n_heads, kv_seq_len, head_dim)
-                V = V.reshape(bsz, n_heads, kv_seq_len, head_dim)
-            pass
-            # Must be contiguous or else results are False!
-            # https://github.com/pytorch/pytorch/issues/112577
-            Q, K, V = Q.contiguous(), K.contiguous(), V.contiguous()
-            # Needs (batch_size, n_heads, seq_len, head_dim)
-            # is_casual and attention_mask must not be both set!
-            A = scaled_dot_product_attention(
-                Q, K, V, attn_mask = attention_mask, is_causal = is_causal
-            )
-            # Go back to (batch_size, seq_len, n_heads, head_dim)
-            A = A.transpose(1, 2).contiguous()
-        pass
+    use_varlen = seq_info is not None and past_key_value is None
```

Comment:
Did you confirm if Llama 3.2 1B Instruct or others have inference not broken in the notebooks? Since we have a different path for inference